std/algorithm/sorting.d

```diff
@@ -4465,3 +4468,260 @@ shapes. Here's a non-trivial example:
     }
     assert(n == 60);
 }
+
+/** This function transform `decimal` value into a value in the factorial number
+system stored in `fac`.
+
+A factorial number is constructed as:
+fac[0] * 0! + fac[1] * 1! + ... fac[20] * 20!
+
+Params:
+    deci = The value to convert into the factorial number system.
+    fac = The array to store the factorial number. The array is of size 21 as
+        `ulong.max` requires 21 digits in the factorial number system.
+Returns:
+    A variable storing the number of digits of the factorial number stored in
+    `fac`.
+
+See_Also:
+    $(HTTP en.wikipedia.org/wiki/Factorial_number_system,
+        Factorial Number System
+    )
+*/
+size_t decimalToFactorial(ulong decimal, ref ubyte[21] fac)
+        @safe pure nothrow @nogc
+{
+    import std.algorithm.mutation : reverse;
+    size_t idx;
+
+    for (ulong i = 1; decimal != 0; ++i)
+    {
+        fac[idx++] = cast(ubyte)(decimal % i);
+        decimal /= i;
```

Comment:
This is sufficiently complex to confuse the compiler into generate two `DIV` instructions when it only needs one. Changing the code to

```d
for (ulong i = 1; decimal != 0; ++i)
{
    auto temp = decimal % i;
    decimal /= i;
    fac[idx++] = cast(ubyte)(temp);
}
```

generates this with GDC

```asm
  mov rax, rdi
  mov edx, 0
  div rcx
  mov rdi, rax
  cmp rcx, 22
  je .L23
  mov BYTE PTR -1[rsi+rcx], dl
```

vs

```asm
  mov rax, rdi
  mov edx, 0
  div rcx
  cmp rcx, 22
  je .L23
  mov BYTE PTR -1[rsi+rcx], dl
  mov rax, rdi
  mov edx, 0
  div rcx
  mov rdi, rax
```

Reply:
changed